core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java

```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
     protected boolean accumulateExceptions() {
         return true;
     }
+
+    private class FailOverToTasksIndexListener implements ActionListener<ListTasksResponse> {
+        private final Task task;
+        private final ListTasksRequest request;
+        private final ActionListener<ListTasksResponse> delegate;
+
+        public FailOverToTasksIndexListener(Task task, ListTasksRequest request, ActionListener<ListTasksResponse> delegate) {
+            this.task = task;
+            this.request = request;
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onResponse(ListTasksResponse listTasksResponse) {
+            if (listTasksResponse.getTasks().isEmpty()) {
+                startGet();
+            } else {
+                delegate.onResponse(listTasksResponse);
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            if (e instanceof NoSuchNodeException) {
+                startGet();
+            }
+            delegate.onFailure(e);
```

Comment:
That looks like a race condition. I think it's also an important enough use case that is might be worth adding a test for.